add query() and query_table() functions

chrisiou opened the pull request:
This PR adds two functions as discussed [here](https://github.com/duckdblabs/duckdb-internal/issues/1026).

```
# Executes a select statement
TABLE_REF query(VARCHAR qstr)

# Parses and evaluates a table or a union of an arbitrary list of tables
TABLE_REF query_table(VARCHAR tbl_name)
TABLE_REF query_table(LIST<VARCHAR> tbl_names [, BOOLEAN by_name])
```
The `query_table()` implements the `union all` function as commented [below](https://github.com/duckdb/duckdb/pull/10586#issuecomment-1951571760) with the difference that it does not allow the execution of expressions other than querying tables.
The optional parameter `by_name` defaults to `false`. When set to `true`, it performs the union operation `BY NAME` (as `union_all` does).

chrisiou committed:
add query() function

carlopi reviewed (approved):
LGTM!

I can think of various follow ups (like using this as backed for json_exeucte_sql), but this is great!

Alex-Monahan commented:
I'm really excited about this!

Will this work when dynamically creating the SQL to be executed? Would this work, for example? It works with the `json_execute_serialized_sql(json_serialize_sql(` workaround. Maybe it is worth adding a few tests even if it is already supported? 

Many thanks!!

```sql
CREATE MACRO union_all(table_list, by_name:='FALSE') AS TABLE (
    FROM json_execute_serialized_sql(
        json_serialize_sql('FROM ' || 
            list_reduce(table_list, 
                (x, y) ->  x || 
                    ' UNION ALL ' || 
                    CASE WHEN by_name THEN 'BY NAME ' ELSE '' END || 
                    'FROM ' || y
            )
        )
    )
);

from union_all(['test', 'test2', 'test3', '(select ''I am a subquery'')'], by_name:='FALSE');
```

samansmink reviewed (commented):
Hey @chrisiou this looks great! I have no comments on your code. 

I think @Alex-Monahan's example works perfectly fine with your code here so maybe adding a similar test is indeed nice?

chrisiou committed:
add an example of using query() to create a dynamic tbl macro

chrisiou commented:
Thank you all for your help and feedback! I also added @Alex-Monahan's test case. 
If you spot any other test cases that should be covered, let me know :)

chrisiou committed:
add query_table() function

chrisiou committed:
add by_name argument in query_table()

chrisiou committed:
change query_table() to get a varchar or a list of varchars as argument

chrisiou commented:
I added the `query_table` function, which has the functionality of the `union_all`(mentioned above), and updated the PR's description with a warning regarding this function.

samansmink commented:
Hey @chrisiou thanks for the update! I think one of the CI jobs failing might have to do with your changes, could you take a look there? https://github.com/duckdb/duckdb/actions/runs/8078725148/job/22076227391?pr=10586

I think you may be missing a `std::move`

chrisiou committed:
change by_name argument from a named parameter to an optional regular one

Alex-Monahan commented:
I am so pumped for this! Syntax directly for `UNION ALL` - love it!!

chrisiou committed:
rename some functions

chrisiou committed:
rmv by_name parameter from query_table(TABLE tbl_name) - it isn't needed

Mytherin commented on test/sql/catalog/function/query_function.test:
```diff
@@ -0,0 +1,240 @@
+1?ch@racter$
+
+query I
+from query_table('(select 17 + 25)');
```
This seems like a problem - I think we should escape input to `query_table`. Otherwise we can't query tables that have these kind of names. For example:

```sql
CREATE TABLE "(select 17 + 25)"(i INT);
INSERT INTO "(select 17 + 25)" VALUES (100);
SELECT * FROM "(select 17 + 25)";
┌───────┐
│   i   │
│ int32 │
├───────┤
│   100 │
└───────┘
```

Could we add this test and make sure this works using special table names as well?

Mytherin replied:
No, `query_table` should only allow querying of tables

Mytherin commented on src/function/table/query_function.cpp:
```diff
@@ -0,0 +1,75 @@
+	                     ? "BY NAME "
+	                     : ""; // 'by_name' variable defaults to false
+	if (input.inputs[0].type().id() == LogicalTypeId::VARCHAR) {
+		query += "FROM " + input.inputs[0].ToString();
```
Could we use `KeywordHelper::WriteOptionallyQuoted()` here to correctly escape the table name?

Mytherin reviewed (commented):
Thanks for the PR! Looks good - I've left some comments

chrisiou committed:
change query_table function to query only tables

samansmink reviewed (approved):
LGTM!

Mytherin commented:
Thanks!
